crates/kitsune_p2p/dht/src/arq/arq_set.rs

```diff
@@ -305,6 +304,95 @@ mod tests {
         );
     }
 
+    #[test]
+    fn arq_set_is_union() {
+        let dim = SpaceDimension::standard();
+        let strat = ArqStrat::default();
+
+        let start_a = 10_000;
+        let len_a = 30_000;
+        let arq_a =
+            Arq::from_start_and_half_len_approximate(dim, &strat, start_a.into(), len_a / 2);
+
+        let start_b = 10_000_000;
+        let len_b = 20_000;
+        let arq_b =
+            Arq::from_start_and_half_len_approximate(dim, &strat, start_b.into(), len_b / 2);
+
+        let arq_set = ArqSet::new(vec![arq_a.to_bounds_std(), arq_b.to_bounds_std()]);
+        let arc_set = arq_set.to_dht_arc_set_std();
+
+        // Before first interval
+        {
+            let agent_arc_before_a =
+                Arq::from_start_and_half_len_approximate(dim, &strat, 100.into(), 100);
+            let interval = DhtArcRange::from(agent_arc_before_a.to_dht_arc_std());
+
+            assert!(!arc_set.overlap(&interval.into()));
+        }
+
+        // Overlaps with start of first interval
+        {
+            let agent_arc_overlap_start_a = Arq::from_start_and_half_len_approximate(
+                dim,
+                &strat,
+                (start_a - 1_000).into(),
+                1_500,
+            );
+            let interval = DhtArcRange::from(agent_arc_overlap_start_a.to_dht_arc_std());
+
+            assert!(arc_set.overlap(&interval.into()));
+        }
```

Comment:
`100` and `1500` are both sub-quantum values, so it's misleading to use different values which round to the same thing (4096). I think this could be a more useful test if it used larger scales which comprise many quanta.

Reply:
I understand the point that smaller networks will have far larger arcs than this, yes I agree. I don't see why it follows that unit testing should avoid testing at this scale. My aim here was to pick numbers that I can easily read off and see the logic. If i use calculations or much larger numbers then I'm not able to do that.

Using sub-quantum values helped reveal an issue with small values where they are collapsed to 0 rather than expanded to 1 quanta. Even if the code I found the problem in is test-only, I still think that helps justify this test.

I'm also thinking that testing at an arbitrary scale like this should generalise. If this works, then larger arcs and more arcs should also work.

Reply:
I agree with your reasoning, but in at least one way this is a special case that doesn't generalize:

The arc resizing algo always produces arcs that are between 8 and 16 segments in size, and the minimum segment size is of course 1 quantum. So, for the entire arc to be only 1 quantum long means we have a single-segment arc, which is not typical. Technically the algo should go all the way down to 1 quantum segment if it needs to, but that is the only time an arc will contain only 1 segment.

It's certainly OK to test this, but I think the usefulness is more in that it *is* an extreme edge case. For testing virtually all other cases, we at least need arcs that are 8-16 segments long.

Reply:
Understood, yes that's fair that in the context that these types are used these tests are not realistic. I was trying to understand the pieces in isolation and I'm happy with these being an extreme case in that context.